### PR TITLE
GEODE-5281: replicate failure and fix bug

### PR DESCRIPTION
**Pull request by balesh2**
- replicate the bad timing using a DUnit test
- fix the bug by only doing one lookup of MBeans
- add more tests
- fix potential NPE in waiting for regions to be ready on servers

Co-authored-by: Kenneth Howe <khowe@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by balesh2** [comment]
Failures in Travic CI and "Test" job of precheckin are due to a known failure that has been fixed on develop since this PR was made.

File: geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java [hunk 1]
```diff
@@ -100,6 +100,10 @@ public FederatingManager(MBeanJMXAdapter jmxAdapter, ManagementResourceRepo repo
     this.messenger = new MemberMessenger(jmxAdapter, repo, system);
   }
 
+  void setProxyFactory(MBeanProxyFactory newProxyFactory) {
```
Comment (pdxrunner) [inline geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java]: I'm still not real happy that we had to add this method to product code be able to test to allow mocking/spying MBeanProxyFactory for unit testing. I think that we should add the `@TestingOnly` annotation to the method to highlight that it is only intended to be used in tests.

Comment (balesh2) [reply geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java]: Good idea. I'll add that now. I don't think there is a way around adding this method for the test though.

Comment (pdxrunner) [reply geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java]: I agree that it's needed for exposing the original problem in a deterministic test. At least adding the annotation won't affect the test runs.

**Comment by pdxrunner** [comment]
UI and acceptance test failures are not related to this change.

**Pull request merged**